src/main/scala/edu/ie3/simona/agent/grid/congestion/VoltageRange.scala

```diff
@@ -0,0 +1,318 @@
+      VoltageRange(
+        deltaPlus,
+        deltaMinus,
+        suggestion.asPu,
+      )
+    } else {
+      // the voltage in this range is fine, set the suggested voltage change to zero
+      VoltageRange(
+        deltaPlus,
+        deltaMinus,
+        0.asPu,
+      )
+    }
+  }
+
+  /** Method to calculate the possible range of voltage changes.
+    *
+    * @param powerFlowResultEvent
+    *   Results from simulating the grid.
+    * @param voltageLimits
+    *   Voltage limits.
+    * @param gridComponents
+    *   All components of the grid.
+    * @param inferiorData
+    *   Map: inferior grid to [[VoltageRange]] and [[TransformerTappingModel]].
+    * @return
+    *   A [[VoltageRange]].
+    */
+  def apply(
+      powerFlowResultEvent: PowerFlowResultEvent,
+      voltageLimits: VoltageLimits,
+      gridComponents: GridComponents,
+      inferiorData: Map[ActorRef[
+        GridAgent.Message
+      ], (VoltageRange, Set[TransformerTapping])],
+      subnetNo: Int,
+  ): VoltageRange = {
+    // filter nodes in subnet
+    val nodesInSubnet =
+      gridComponents.nodes.filter(_.subnet == subnetNo).map(_.uuid)
+
+    // calculate voltage range
+    val nodeResMap = powerFlowResultEvent.nodeResults
+      .filter(res => nodesInSubnet.contains(res.getInputModel))
+      .map(res => res.getInputModel -> res.getvMag())
+      .toMap
+    val minVoltage = nodeResMap
+      .minByOption(_._2)
+      .getOrElse(throw new CriticalFailureException(s"No node result found!"))
+    val maxVoltage = nodeResMap
+      .maxByOption(_._2)
+      .getOrElse(throw new CriticalFailureException(s"No node result found!"))
+
+    // build initial range
+    val range = VoltageRange(
+      voltageLimits.vMax.subtract(maxVoltage._2),
+      voltageLimits.vMin.subtract(minVoltage._2),
+    )
+
+    // updating the voltage range prevent or cure line congestions
+    val deltaV = calculatePossibleVoltageDeltaForLines(
+      nodeResMap,
+      powerFlowResultEvent.lineResults,
+      gridComponents,
+    )
+    val updatedRange = range.updateWithLineDelta(deltaV)
+
+    if inferiorData.isEmpty then {
+      // if there are no inferior grids, return the voltage range
+      updatedRange
+    } else {
+      // if there are inferior grids, update the voltage range
+      updatedRange.updateWithInferiorRanges(inferiorData)
+    }
+  }
+
+  /** Method to calculate a voltage delta for the given line currents. <p> - If
+    * there is a line congestion, increasing the voltage by the returned delta
+    * should mitigate them. <p> - If there is no line congestion, the returned
+    * voltage shows the possible voltage decrease. <p> - Formula: V * I = (V +
+    * deltaV) * (I + deltaI).
+    *
+    * @param nodeResults
+    *   Node voltages.
+    * @param lineResults
+    *   Line currents.
+    * @param gridComponents
+    *   Information of components.
+    * @return
+    *   A voltage delta.
+    */
+  def calculatePossibleVoltageDeltaForLines(
```

Comment:
```suggestion
  def calculateVoltageDeltaFromLineCurrent(
```
to me it's more like this, but I'm open for alternatives